paimon-python/pypaimon/write/partial_column_write.py

```diff
@@ -0,0 +1,187 @@
+        for row_id in row_ids:
+            first_row_id = self._floor_binary_search(self.first_row_ids, row_id)
+            first_row_id_values.append(first_row_id)
+
+        # Add first_row_id column to the table
+        first_row_id_array = pa.array(first_row_id_values, type=pa.int64())
+        return data.append_column(self.FIRST_ROW_ID_COLUMN, first_row_id_array)
+
+    def _floor_binary_search(self, sorted_seq: List[int], value: int) -> int:
+        """Binary search to find the floor value in sorted sequence."""
+        if not sorted_seq:
+            raise ValueError("The input sorted sequence is empty.")
+
+        idx = bisect.bisect_right(sorted_seq, value) - 1
+        if idx < 0:
+            raise ValueError(f"Value {value} is less than the first element in the sorted sequence.")
+
+        return sorted_seq[idx]
+
+    def _write_by_first_row_id(self, data: pa.Table, column_names: List[str]):
+        """Write data grouped by first_row_id."""
+        # Extract unique first_row_id values
+        first_row_id_array = data[self.FIRST_ROW_ID_COLUMN]
+        unique_first_row_ids = pc.unique(first_row_id_array).to_pylist()
+
+        for first_row_id in unique_first_row_ids:
+            # Filter rows for this first_row_id
+            mask = pc.equal(first_row_id_array, first_row_id)
+            group_data = data.filter(mask)
+
+            # Get partition for this first_row_id
+            partition = self._find_partition_by_first_row_id(first_row_id)
+
+            if partition is None:
+                raise ValueError(f"No existing file found for first_row_id {first_row_id}")
+
+            # Write this group
+            self._write_group(partition, first_row_id, group_data, column_names)
+
+    def _find_partition_by_first_row_id(self, first_row_id: int) -> Optional[GenericRow]:
+        """Find the partition for a given first_row_id using pre-built partition map."""
+        return self.first_row_id_to_partition_map.get(first_row_id)
+
+    def _write_group(self, partition: GenericRow, first_row_id: int,
+                     data: pa.Table, column_names: List[str]):
+        """Write a group of data with the same first_row_id."""
+
+        # Create a file store write for this partition
+        file_store_write = FileStoreWrite(self.table, self.commit_user)
+
+        # Set write columns to only update specific columns
+        # Note: _ROW_ID is metadata column, not part of schema
+        write_cols = column_names
+        file_store_write.write_cols = write_cols
+
+        # Convert partition to tuple for hashing
+        partition_tuple = tuple(partition.values)
+
+        # Write data - convert Table to RecordBatch
+        data_to_write = data.select(write_cols)
+        for batch in data_to_write.to_batches():
+            file_store_write.write(partition_tuple, 0, batch)
+
+        # Prepare commit and assign first_row_id
+        commit_messages = file_store_write.prepare_commit(BATCH_COMMIT_IDENTIFIER)
```

Comment:
We should set sequence number to define the version.

Reply:
sequence number will be set in commit stage.